File: easy/linked_list/remove_duplicates/remove_duplicates.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
        return

    def push(self, val):
        new_node = Node(val)
        curr = self
        while curr.next != None:
            curr = curr.next
        curr.next = new_node
        return

    def print(self):
        curr = self
        while curr != None:
            print(curr.val, end=" -> ")
            curr = curr.next
        print()


class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None or head.next == None:
            return head

        prev, curr = head, head.next
        while curr != None:
            if prev.val == curr.val:
                prev.next = curr.next
            else:
                prev = curr
            curr = curr.next
        return head


s = Solution()

print("Solution 1")
head = Node(1)
head.push(1)
head.push(2)
head.print()
noDuplicates = s.deleteDuplicates(head)
noDuplicates.print()

print("Solution 2")
head = Node(1)
head.push(1)
head.push(2)
head.push(3)
head.push(3)
head.print()
noDuplicates = s.deleteDuplicates(head)
noDuplicates.print()

File: easy/linked_list/remove_duplicates/remove_duplicates.go
package main

import (
	"fmt"
)

type Node struct {
	Val  int
	Next *Node
}

func (n *Node) push(val int) {
	newNode := &Node{Val: val}
	head := n
	for head.Next != nil {
		head = head.Next
	}
	head.Next = newNode
	return
}

func (n *Node) print() {
	head := n
	for head != nil {
		fmt.Printf("%d -> ", head.Val)
		head = head.Next
	}
	fmt.Println()
	return
}

func deleteDuplicates(head *Node) *Node {
	if head == nil || head.Next == nil {
		return head
	}

	prev, curr := head, head.Next
	for curr != nil {
		if prev.Val == curr.Val {
			prev.Next = curr.Next
		} else {
			prev = curr
		}
		curr = curr.Next
	}
	return head
}

func main() {
	fmt.Println("Solution 1")
	head := &Node{Val: 1}
	head.push(1)
	head.push(2)
	head.print()
	noDuplicates := deleteDuplicates(head)
	noDuplicates.print()

	fmt.Println("Solution 2")
	head = &Node{Val: 1}
	head.push(1)
	head.push(2)
	head.push(3)
	head.push(3)
	head.print()
	noDuplicates = deleteDuplicates(head)
	noDuplicates.print()
}
